### ISSUE TITLE
Memory used increasing slowly

### ISSUE THREAD

**Issue by marcosmartinez7** [ethereum/go-ethereum#17450]
#### System information

Geth version: `1.8.12 stable`
OS & Version: Linux 16.04

#### Expected behaviour

Memory usage stays constant

#### Actual behaviour

Hi @karalabe, 

im running the node without the rpcapis, the node started 3 days ago using 1.9% of my RAM (8gb). Now is consuming 2.3% and it keeps increasing slowly (like 10mb / h). 

I ran the node without specifing the --cache flag, so i assume it is using 1gb. 

Is this something that i must worry about or maybe is related with the garbage collection? 

#### Steps to reproduce the behaviour

I ran the node with this command:

geth --datadir e1/ --syncmode 'full' --port 30357 --rpc --rpcport 8545 --rpccorsdomain '*' --rpcaddr 'server_ip' --ws --wsaddr "server_ip" --wsorigins "some_ip" --wsport 9583 --wsapi 'db,eth,net,web3,txpool,miner'  --networkid 21 --gasprice '1'

**Comment by marcosmartinez7** [comment]
Hi, yes, but i need to hit the smart contract from Metamask from any origin. Is there any way to accomplish that without opening up the HTTP interface?

About memory, im not doing any request, the memory usage is increasing without hitting the smart contract. I tought that it might be related to the garbage collector, maybe isnt executing...

Is there any information that i can provide in order to identify the possible leak?

**Comment by marcosmartinez7** [comment]
Recently i get attached to a console and send 400 transactions (hitting a Smart contract) in a for loop. 

The memory used increased 400mb

any idea of what can i research in order to check what is causing this?

**Comment by marcosmartinez7** [comment]
I have tried on another chain that doesnt expose RPC endpoint, sending 5000 transactions from geth console. 

Memory started at 1140 MB and after 5000 transactions grows to 1550 MB, so geth takes 400 MB in order to process those transactions. 

Since the block time is 15 seconds, it will take a while to confirm those transactions, so .. it is normal to stay into 1550MB for a while ?

**Comment by cdljsj** [comment]
The problem gets worse in 1.8.20, I have to restart geth node every half day due to the high memory usage.

**Comment by hapsody** [comment]
Same problem here 
I used --cache flag (--cache "64") but problem is still occured

Version: 1.8.16-stable
Git Commit: 477eb0933b9529f7deeccc233cc815fe34a8ea56
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.11
Operating System: linux
GOPATH=
GOROOT=/home/travis/.gimme/versions/go1.11.linux.amd64

**Comment by hadv** [comment]
I have the same problem. It seems there're memory leak somewhere. And it seems that if node is mining and open the rpc endpoint have no problem. 

If the node is not mining then the memory increase regularly as @marcosmartinez7 report when sending thousand of tnx continuously via rpc endpoint.

**Comment by marcosmartinez7** [comment]
On 1.8.18 version i havent experimented anymore this problem, take in care that geth has in memory some information that is written to disk on each epoch, and an epoch is about 20-30k blocks. 

So if youre using less than 4GB of RAM i think this can happen.

**Comment by hadv** [comment]
Thank you for your information @marcosmartinez7 

By the way, I still can reproduce the issue on 1.8.20 & 1.8.21 with 16GB of RAM by sending 10 thousands of tnx. 

And one thing very strange is mining node dont have the issue on the same computing configuration. So I think there's must be memory leak somewhere.

**Comment by hadv** [comment]
Thank you for your information @marcosmartinez7 

By the way, I still can reproduce the issue on 1.8.20 & 1.8.21 with 16GB of RAM by sending 10 thousands of tnx (2000 txs/block sealing & txpool always fill up with ~10,000 txs)

And one thing very strange is mining node with public RPC endpoint don't have the issue on the same computing configuration. So I think there's must be memory leak somewhere.

**mining**
![image](https://user-images.githubusercontent.com/941411/51849466-5dd45e00-2352-11e9-8d1e-bde459ffb98e.png)

**un-mining** (**memory leak**)

![image](https://user-images.githubusercontent.com/941411/51849572-9411dd80-2352-11e9-83ba-3feb5f194b8c.png)

**Comment by hadv** [comment]
cc: @karalabe @fjl 
goroutine leak on unmining RPC node

![image](https://user-images.githubusercontent.com/941411/51881076-5d68b100-23ac-11e9-97ca-115fa89a807f.png)

mining node (no leak)

![image](https://user-images.githubusercontent.com/941411/51881086-6a85a000-23ac-11e9-94d1-89f61040936a.png)

**Comment by hadv** [comment]
the goroutine leak might be the case of feed.Send() blocking issue reported on https://github.com/ethereum/go-ethereum/issues/18021

**Comment by fjl** [comment]
The issue is not that Feed.Send() is blocking, it's that the send to the feed happens in a goroutine. Please provide a longer stack trace so we can see which part of the system is trying to send on the feed.

**Comment by hadv** [comment]
> The issue is not that Feed.Send() is blocking, it's that the send to the feed happens in a background goroutine. 

I think that the issue happen as @liuzhijun23 figure out in the 18021 when the for loop is blocked in feed.Send() then another call will be stuck at line 133 `<-f.sendLock` because at which the `f.sendLock` is empty

> Please provide a longer stack trace so we can see which part of the system is trying to send on the feed.

I get the trace from debug pprof and that's it 

```
goroutine profile: total 49352
48057 @ 0x44384b 0x4438f3 0x41b15e 0x41ae4b 0x6edc6f 0x472051
#	0x6edc6e	github.com/ethereum/go-ethereum/event.(*Feed).Send+0x12e	/home/admin/gonex/build/_workspace/src/github.com/ethereum/go-ethereum/event/feed.go:133
```

**Comment by hadv** [comment]
@fjl  **a clue**: on no-mining node, the `TrySend()` return **_false_** very frequency then make more and more goroutine leak overtime. On the mining node, `TrySend()` always succeed 

```feed.go
			if cases[i].Chan.TrySend(rvalue) {
				nsent++
				cases = cases.deactivate(i)
				i--
			} 
```

**Comment by hadv** [comment]
I'm investigating more but I think because even node running without `--mine` option but the miner and worker still was create and subscribe to the txpool's feed via below code

```
	// Subscribe NewTxsEvent for tx pool
	worker.txsSub = eth.TxPool().SubscribeNewTxsEvent(worker.txsCh)
```

but when txpool call `txFeed.Send()` to send `NewTxsEvent` to all subscriber including miner/worker but because node is no-mining then the sending() to miner/worker is false that make goroutine leak.

**Comment by hadv** [comment]
@holiman hope this will help

https://drive.google.com/open?id=1xK6F95bmLsleuvM9qMQaH2Y7K9Gk2ZFD

**Comment by holiman** [comment]
You've got `203` threads stuck on 
```
1: semacquire [Created by http.(*Server).Serve @ server.go:2851]
    sync       sema.go:71                      runtime_SemacquireMutex(*uint32(#1648), bool(#6018))
    sync       rwmutex.go:50                   (*RWMutex).RLock(*RWMutex(#1647))
    miner      worker.go:252                   (*worker).pending(#1646, 0, 0)
    miner      miner.go:155                    (*Miner).Pending(#1676, #28503, #129)
    eth        api_backend.go:92               (*EthAPIBackend).StateAndHeaderByNumber(#1920, #27, #16012, 0xfffffffffffffffe, #45, 0, 0, 0x5208)
    ethapi     api.go:700                      (*PublicBlockChainAPI).doCall(#1121, #27, #16012, #28616, #28747, #244, #22014, 0x551ae, 0, 0, ...)
    ethapi     api.go:791                      (*PublicBlockChainAPI).EstimateGas.func1(0x551ae, #26)
    ethapi     api.go:800                      (*PublicBlockChainAPI).EstimateGas(#1121, #27, #16012, #28616, #28747, #244, #22014, 0x551ae, 0, 0, ...)
    reflect    value.go:447                    Value.call(string(#2177, len=824635791808), []Value(0x13 len=16418845 cap=4), #2846, 0x3, 0x4, 0x0, #2846, ...)
    reflect    value.go:308                    Value.Call([]Value(#2177 len=824635791808 cap=19), #2846, 0x3, 0x4, 0x1, 0x1, 0x0)
    rpc        server.go:309                   (*Server).handle(#3676, #27, #16012, #37, #23615, #2845, #23616, 0, 0xe630e0)
    rpc        server.go:330                   (*Server).exec(#3676, #27, #16012, #37, #23615, #2845)
    rpc        server.go:192                   (*Server).serveRequest(#3676, #29, #27953, #37, #23615, 0xfb1801, 0x1, 0, 0)
    rpc        server.go:223                   (*Server).ServeSingleRequest(#3676, #29, #27953, #37, #23615, 0x1)
    rpc        http.go:257                     (*Server).ServeHTTP(#3676, #24, #12949, #14736)
    cors       cors.go:190                     (*Cors).Handler.func1(#24, #12949, #14736)
    http       server.go:1964                  HandlerFunc.ServeHTTP(ResponseWriter(#3660), *Request(#12949), #14736)
    rpc        http.go:324                     (*virtualHostHandler).ServeHTTP(#3661, #24, #12949, #14736)
    http       server.go:2741                  serverHandler.ServeHTTP(ResponseWriter(#3680), *Request(#12949), #14736)
```

If you are batch-adding thousands of transactions, and doing an `estimategas` on each and every one, it will be quite resource intensive for the node. They will each be competing for the lock to obtain a particular state. 
I guess a more batch-friendly method could be used, that would take the `pending` state, and reuse it for every tx in the batch. But however we do it, it'll be messy -- e.g. do we want to apply transactions on top of eachother or reset the state again after each?

**Comment by hadv** [comment]
@holiman Can you please explain us why only not-mining node need to run below code? That's might be the reason why only not-mining node face the goroutine leak issue, right? Thank you!

```
		case ev := <-w.txsCh:
			// Apply transactions to the pending state if we're not mining.
			//
			// Note all transactions received may not be continuous with transactions
			// already included in the current mining block. These transactions will
			// be automatically eliminated.
			if !w.isRunning() && w.current != nil {
				w.mu.RLock()
				coinbase := w.coinbase
				w.mu.RUnlock()

				txs := make(map[common.Address]types.Transactions)
				for _, tx := range ev.Txs {
					acc, _ := types.Sender(w.current.signer, tx)
					txs[acc] = append(txs[acc], tx)
				}
				txset := types.NewTransactionsByPriceAndNonce(w.current.signer, txs)
				w.commitTransactions(txset, coinbase, nil)
				w.updateSnapshot()
			} else {
```

**Comment by holiman** [comment]
I don't know yet.. However, there appears to be ~10K routines spawned by `promoteExecutables`, that are waiting on the lock in `feed.go`. One is busy in the loop. 

I think an underlying problem is that a better model for the transaction handling would be to use active objects (one thread/routine) which receives data, instead of each sender spawning it's own goroutine (https://github.com/ethereum/go-ethereum/blob/master/core/tx_pool.go#L1000) . Btw, @hadv , I don't know what code you're running, but the line numbers from your stack does not match up with what's on master now. 

I'm not sure if there are any simple solutions to this ticket, since IMO it would probably require a non-trivial rewrite of tx pool internals.

**Comment by hadv** [comment]
okay, thank you for your information. For the code I'm adding some log to figure out the issue then the line of code might be difference with the master but the logic are the same though.

**Comment by bishaoqing** [comment]
> @holiman Can you please explain us why only not-mining node need to run below code? That's might be the reason why only not-mining node face the goroutine leak issue, right? Thank you!
> 
> ```go
> 		case ev := <-w.txsCh:
> 			// Apply transactions to the pending state if we're not mining.
> 			//
> 			// Note all transactions received may not be continuous with transactions
> 			// already included in the current mining block. These transactions will
> 			// be automatically eliminated.
> 			if !w.isRunning() && w.current != nil {
> 				w.mu.RLock()
> 				coinbase := w.coinbase
> 				w.mu.RUnlock()
> 
> 				txs := make(map[common.Address]types.Transactions)
> 				for _, tx := range ev.Txs {
> 					acc, _ := types.Sender(w.current.signer, tx)
> 					txs[acc] = append(txs[acc], tx)
> 				}
> 				txset := types.NewTransactionsByPriceAndNonce(w.current.signer, txs)
> 				w.commitTransactions(txset, coinbase, nil)
> 				w.updateSnapshot()
> 			} else {
> ```

it might  want to refresh the pending state,so that the rpc-client can get the latest pending information,for example, nonce of account, and users don't have to maintain the information themselves

**Comment by fjl** [comment]
I think a simple fix to try would be removing the `go` on this line https://github.com/ethereum/go-ethereum/blob/master/core/tx_pool.go#L1000.

**Comment by bishaoqing** [comment]
Mining-node will execute pending-txs by commitNewWork,but none-minging isn't,so i guess that none-minging node do this for keep the pending-state up to date

**Comment by hadv** [comment]
> Removing the goroutine would just mean that the callers of those methods need to wait until the events have been delivered.

Yeah, I'm afraid that in case Feed.Send() is blocked then whole txpool is blocked also

**Comment by hadv** [comment]
by the way, I think we have enough information for the issue then could you remove the label `need:more-information`? @fjl Thanks!

**Comment by jbunce** [comment]
Here's some data from a `geth` running `1.8.22` over the past week: https://imgur.com/a/2LJpY8U

The first drop was when the node was upgraded from `1.8.21` to `1.8.22`. The second one was from where the node was restarted with mining enabled.

**Comment by hadv** [comment]
> by the way, I think we have enough information for the issue then could you remove the label `need:more-information`? @fjl Thanks!

@karalabe @holiman @fjl Will we have any update on this issue in short term?

**Comment by hadv** [comment]
@fjl any more information do you need for this issue?

**Comment by hadv** [comment]
> This issue has been automatically closed because there has been no response to our request for more information from the original author. With only the information that is currently in the issue, we don't have enough information to take action. Please reach out if you have more relevant information or answers to our questions so that we can investigate further.

@karalabe @fjl @holiman We already provide the details information. Please remove the inappropriate label and re-open this issue. Thank you!

**Comment by hadv** [comment]
open new issue #19192
